lib/Module.js

```diff
@@ -171,18 +164,45 @@ class Module extends DependenciesBlock {
 		return false;
 	}
 
+	// Remove oldChunk from and add newChunks to the module's reasons.
+	// Use a hash map to avoid adding duplicate chunks. This approach is used
+	// rather than indexOf for performance reasons.
 	rewriteChunkInReasons(oldChunk, newChunks) {
-		this.reasons.forEach(r => {
+		// create the newChunksMap outside the reasons loop to avoid duplicating
+		// the work.
+		const newChunkMap = {};
+		newChunks.forEach(chunk => {
+			newChunkMap[chunk.debugId] = chunk;
+		});
+
+		for(let reasonIndex = 0; reasonIndex < this.reasons.length; reasonIndex++) {
+			const r = this.reasons[reasonIndex];
 			if(!r.chunks) {
 				if(!r.module._chunks.has(oldChunk))
-					return;
+					continue;
 				r.chunks = Array.from(r.module._chunks);
 			}
-			r.chunks = r.chunks.reduce((arr, c) => {
-				addToSet(arr, c !== oldChunk ? [c] : newChunks);
-				return arr;
-			}, []);
-		});
+
+			const chunksMap = {};
+			for(let chunkIndex = 0; chunkIndex < r.chunks.length; chunkIndex++) {
+				const chunk = r.chunks[chunkIndex];
+				if(chunk !== oldChunk) {
+					chunksMap[chunk.debugId] = chunk;
+				}
+			}
+			// override any matching chunkIds
+			Object.assign(chunksMap, newChunkMap);
+
+			// Convert each entry in the chunksMap into a value in an array.
+			// Creating a fixed length array seems to be the fastest method of doing this.
+			// https://jsperf.com/js-for-vs-map
+			const keys = Object.keys(chunksMap);
+			r.chunks = new Array(keys.length);
+			for(let keyIndex = 0; keyIndex < keys.length; keyIndex++) {
+				const chunk = chunksMap[keys[keyIndex]];
+				r.chunks[keyIndex] = chunk;
```

Comment:
I can refactor this to be whatever, but this is what seemed to be the fastest when I was profiling locally.